### PR TITLE
Add Towncrier changelog fragments

### PR DESCRIPTION
**Pull request by ganeshrn**
##### SUMMARY

*  Add changelog fragments

##### ISSUE TYPE

- Docs Pull Request

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -74,3 +74,45 @@ profile = "True"
...
+
+[[tool.towncrier.type]]
+    directory = "doc"
+    name = "Documentation"
+    showcontent = true
+
+[[tool.towncrier.type]]
+    directory = "misc"
+    name = "Miscellaneous"
+    showcontent = true
```
Comment (webknjaz) [inline pyproject.toml]: This breaks the assumption of having another section the last. See the comment above, it should be relocated to keep the scm setup last

File: docs/changelog/fragments/10.feature.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Add proxycommand support for session and
+update session execption to LibsshSessionException - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog/fragments/10.feature.rst]: JFYI the `:user:` role doesn't work yet, I'll need to add it to extlinks in conf.py, but ignore it for now. You can use this but sphinx build will fail (for now).

File: docs/changelog/fragments/113.bugfix.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Fix double close issue, ddd logic to free
+the channel allocated memory within
+`__dealloc__`  - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog/fragments/113.bugfix.rst]: Since we use RST syntax, markdown assumptions won't work as you expect. Of course, I've set the default role to "any" but it'll only work if sphinx finds something to link against. Currently, Cython docs are not added to intersphinx so the only thing you can do is to mark it as inline code:
```suggestion
``__dealloc__`` - by :user:`ganeshrn`
```

File: docs/changelog/fragments/58.bugfix.rst [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Fix sftp file get issue on py2
+The file write() method returns None on py2
```
Comment (webknjaz) [inline docs/changelog/fragments/58.bugfix.rst]: I think that most of the code references should be marked as inline code
```suggestion
The file ``write()`` method returns None on py2
```

File: docs/changelog/README.rst [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+    Added sftp get functionality to fetch file from remote host  - by :user:`amolkahat`
+
+file ``571.bugfix.rst``::
+
+    Fix double close issue, ddd logic to free
+    the channel allocated memory within
+    `__dealloc__`  - by :user:`ganeshrn`
+
+.. see pyproject.toml for all available categories  (tool.towncrier.type)
```
Comment (webknjaz) [inline docs/changelog/README.rst]: This is a code comment, it won't show up in the render but it should (as README files show up in GitHub UI when the folder view is open)

File: docs/changelog/README.rst [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+file ``544.feature.rst``::
+
+    Added sftp get functionality to fetch file from remote host  - by :user:`amolkahat`
+
+file ``571.bugfix.rst``::
+
+    Fix double close issue, ddd logic to free
+    the channel allocated memory within
+    `__dealloc__`  - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog/README.rst]: Let's advertise proper inline code markings
```suggestion
    ``__dealloc__`` - by :user:`ganeshrn`
```

File: docs/changelog/README.rst [hunk 7]
```diff
@@ -0,0 +1,18 @@
+.. examples for changelog entries adding to your Pull Requests
```
Comment (webknjaz) [inline docs/changelog/README.rst]: this won't show up anywhere as long as `.. ` is in front of it

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: docs/changelog/README.rst (hunk 6 above)
Comment (webknjaz) [reply docs/changelog/README.rst]: Actually, try this instead:
```suggestion
    `__dealloc__() <finalization_method>` - by :user:`ganeshrn`
```
(it'll work once I finish my intersphinx PR)

File: docs/changelog/fragments/113.bugfix.rst (hunk 3 above)
Comment (webknjaz) [reply docs/changelog/fragments/113.bugfix.rst]: this should work later
```suggestion
`__dealloc__() <finalization_method>` - by :user:`ganeshrn`
```

File: pyproject.toml [hunk 8]
```diff
@@ -74,3 +74,45 @@ profile = "True"
 # the end when publishing non-tagged versions.
 [tool.setuptools_scm]
 write_to = "src/pylibsshext/_scm_version.py"
+
+[tool.towncrier]
+package = "tox"
```
Comment (webknjaz) [inline pyproject.toml]: (forgotten copy-paste)

File: docs/changelog/README.rst [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+    Added sftp get functionality to fetch file from remote host  - by :user:`amolkahat`
+
+file ``571.bugfix.rst``::
+
+    Fix double close issue, ddd logic to free
+    the channel allocated memory within
+    `__dealloc__() <finalization_method>`  - by :user:`ganeshrn`
+
+.. Note::
```
Comment (webknjaz) [inline docs/changelog/README.rst]: I believe it shouldn't be capitalized but nevermind, I'm going to have this file included in my PR in a proper state anyway.

**Comment by webknjaz** [comment]
It's a bit better now: https://ansible-pylibssh--114.org.readthedocs.build/en/114/changelog.html#vrelease-unreleased-draft-2020-07-31.

File: docs/changelog-fragments/1.feature.rst [hunk 10]
```diff
@@ -0,0 +1,2 @@
+Add cython extension for libssh client
```
Comment (webknjaz) [inline docs/changelog-fragments/1.feature.rst]: To make a good changelog, the sentence style shouldn't be imperative. They should be written in the past tense.

File: docs/changelog-fragments/1.feature.rst [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
+API's initial commit - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/1.feature.rst]: I've modified the documentation example to use double dash as in `--` instead of `-` (it's turned into an M-dash when rendered).
```suggestion
API's initial commit -- by :user:`ganeshrn`
```

File: docs/changelog-fragments/58.bugfix.rst [hunk 12]
```diff
@@ -0,0 +1,7 @@
+Fix sftp file get issue on py2
+The file ``write()`` method returns None on py2
+is bytes are written to file successfully, whereas
+on py3 it returns total number of bytes written
+to file. Add fix to check for the number of
+bytes written only in the case when ``write()``
+is not returning None - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/58.bugfix.rst]: Things like `None` should be marked as inline-code with double backticks
```suggestion
is not returning ``None`` -- by :user:`ganeshrn`
```

File: docs/changelog-fragments/113.bugfix.rst [hunk 13]
```diff
@@ -0,0 +1,3 @@
+Fix double close issue, ddd logic to free
+the channel allocated memory within
+`__dealloc__() <finalization_method>`  - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/113.bugfix.rst]:
```suggestion
`__dealloc__() <finalization_method>` -- by :user:`ganeshrn`
```

File: docs/changelog-fragments/10.feature.rst [hunk 14]
```diff
@@ -0,0 +1,2 @@
+Add proxycommand support for session and
+update session execption to LibsshSessionException - by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/10.feature.rst]:
```suggestion
update session exeception to ``LibsshSessionException`` - by :user:`ganeshrn`
```

**Review by webknjaz — CHANGES_REQUESTED** [review]
I have added a few inline style comments but they apply to all fragments.
TL;DR Turn the fragments style into the past tense and check the preview @ https://ansible-pylibssh--114.org.readthedocs.build/en/114/changelog.html

Another important thing to consider is that these news fragments are supposed to be targeting the end-users (developers in our case) and so they shouldn't mention any internal implementation details but rather talk about the public API changes and environment considerations. Minor things should just be omitted. The log for cores is in Git anyway.

P.S. I've updated the instructions @ https://ansible-pylibssh.readthedocs.io/en/latest/contributing/guidelines.html#adding-change-notes-with-your-prs too.

File: docs/changelog-fragments/1.feature.rst (hunk 10 above)
Comment (ganeshrn) [reply docs/changelog-fragments/1.feature.rst]: The original PR had links to the PR/issue which seems to be removed. imo the changelog should link to PR/issues for which the change was done.

Comment (webknjaz) [reply docs/changelog-fragments/1.feature.rst]: The links are made from the numbers in fragment names. This hasn't changed. Is this what you mean?

File: docs/_ext/towncrier_draft_ext.py [hunk 15]
```diff
@@ -10,7 +10,7 @@
 
 from sphinx.application import Sphinx
 from sphinx.util.docutils import SphinxDirective
-from sphinx.util.nodes import nodes
+from sphinx.util.nodes import nested_parse_with_titles, nodes
```
Comment (webknjaz) [inline docs/_ext/towncrier_draft_ext.py]: Looks like this PR is out of date. It needs updates.

File: docs/changelog-fragments/1.feature.rst (hunk 10 above)
Comment (ganeshrn) [reply docs/changelog-fragments/1.feature.rst]: Ah, I missed that link. Ignore my earlier comment.

File: docs/changelog-fragments/14.doc.rst [hunk 16]
```diff
@@ -0,0 +1,2 @@
+Update requirements file to replace
```
Comment (webknjaz) [inline docs/changelog-fragments/14.doc.rst]: This is not a doc change. It's rather internal.

Comment (ganeshrn) [reply docs/changelog-fragments/14.doc.rst]: not necessarily ...it is also a user/dev facing change.

Comment (webknjaz) [reply docs/changelog-fragments/14.doc.rst]: End-users don't use this file. Only the core devs of this project do.

Comment (webknjaz) [reply docs/changelog-fragments/14.doc.rst]: FWIW this is not a documentation change, it does not edit any docs.

File: docs/changelog-fragments/113.bugfix.rst [hunk 17]
```diff
@@ -0,0 +1,3 @@
+Fix double close issue, ddd logic to free
+the channel allocated memory within
+`__dealloc__() <finalization_method>`  -- by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/113.bugfix.rst]:
```suggestion
`__dealloc__() <finalization_method>` -- by :user:`ganeshrn`
```

File: docs/changelog-fragments/26.feature.rst [hunk 18]
```diff
@@ -0,0 +1,2 @@
+Added sftp get functionality to fetch file
```
Comment (webknjaz) [inline docs/changelog-fragments/26.feature.rst:1]: This is the only place with a sentence in the past tense. All of the others should be changed to be like this.

File: docs/changelog-fragments/14.doc.rst (hunk 16 above)
Comment (ganeshrn) [reply docs/changelog-fragments/14.doc.rst]: Not only core dev anyone who wants to contribute to the project can use the file, so it should be in changelog. I will add it as misc change

File: docs/changelog-fragments/26.feature.rst (hunk 18 above)
Comment (ganeshrn) [reply docs/changelog-fragments/26.feature.rst:1]: That is because of the PR title.

Comment (ganeshrn) [reply docs/changelog-fragments/26.feature.rst:1]: I don't think other places it requires to be changed and it is fine as it is.

File: docs/changelog-fragments/1.feature.rst (hunk 10 above)
Comment (ganeshrn) [reply docs/changelog-fragments/1.feature.rst]: Fixed

File: docs/changelog-fragments/26.feature.rst (hunk 18 above)
Comment (ganeshrn) [reply docs/changelog-fragments/26.feature.rst:1]: Fixed, I don't have a strong feeling about this but since you think it is the right away to add changelog I am fine with it 👍

Comment (webknjaz) [reply docs/changelog-fragments/26.feature.rst:1]: Well, then the changelog will be horribly inconsistent.

Comment (webknjaz) [reply docs/changelog-fragments/26.feature.rst:1]: Oh, I see you fixed the other places.

Comment (webknjaz) [reply docs/changelog-fragments/26.feature.rst:1]: P.S. I wasn't talking about changing the PR titles, this would be too much unnecessary work. Besides, PR titles should follow Git commits standard anyway.

File: docs/changelog-fragments/25.bugfix.rst [hunk 19]
```diff
@@ -0,0 +1 @@
+Changed pylibssh dir to pylibsshext to avoid ns collision -- by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/25.bugfix.rst]: It's not a bugfix. We just renamed the importable. There was no collision. We wanted the name to be more clear AFAIR. If anything, I'd add it as a feature.

File: docs/changelog-fragments/113.bugfix.rst [hunk 20]
```diff
@@ -0,0 +1,3 @@
+Fixed double close issue, ddd logic to free
```
Comment (webknjaz) [inline docs/changelog-fragments/113.bugfix.rst]:
```suggestion
Fixed double close issue, added logic to free
```

File: docs/changelog-fragments/46.bugfix.rst [hunk 21]
```diff
@@ -0,0 +1,3 @@
+Fixed session timeout issue, the data type
+of timeout is expected by ``ssh_options_set``
+is of type long int -- by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/46.bugfix.rst]:
```suggestion
is of type ``long int`` -- by :user:`ganeshrn`
```

File: docs/changelog-fragments/58.bugfix.rst [hunk 22]
```diff
@@ -0,0 +1,7 @@
+Fixed sftp file get issue on py2
+The file ``write()`` method returns None on py2
```
Comment (webknjaz) [inline docs/changelog-fragments/58.bugfix.rst]:
```suggestion
the file ``write()`` method returns ``None`` on py2
```

File: docs/changelog-fragments/58.bugfix.rst [hunk 23]
```diff
@@ -0,0 +1,7 @@
+Fixed sftp file get issue on py2
```
Comment (webknjaz) [inline docs/changelog-fragments/58.bugfix.rst]:
```suggestion
Fixed sftp file get issue. On py2
```

File: docs/changelog-fragments/58.bugfix.rst [hunk 24]
```diff
@@ -0,0 +1,7 @@
...
+The file ``write()`` method returns None on py2
+is bytes are written to file successfully, whereas
```
Comment (webknjaz) [inline docs/changelog-fragments/58.bugfix.rst]:
```suggestion
if bytes are written to file successfully, whereas
```

File: docs/changelog-fragments/58.bugfix.rst [hunk 25]
```diff
@@ -0,0 +1,7 @@
...
+on py3 it returns total number of bytes written
+to file. Add fix to check for the number of
```
Comment (webknjaz) [inline docs/changelog-fragments/58.bugfix.rst]:
```suggestion
to file. Added a fix to check for the number of
```

File: docs/changelog-fragments/58.bugfix.rst [hunk 26]
```diff
@@ -0,0 +1,7 @@
...
+bytes written only in the case when ``write()``
+is not returning ``None`` -- by :user:`ganeshrn`
```
Comment (webknjaz) [inline docs/changelog-fragments/58.bugfix.rst]:
```suggestion
does not return ``None`` -- by :user:`ganeshrn`
```

**Review by webknjaz — APPROVED** [review]
